wasmdome/src/main.rs

```diff
@@ -27,81 +36,146 @@ struct CliCommand {
-    let homedir = dirs::home_dir();
-    if homedir.is_none() {
-        return Err("Cannot locate home directory".into());
-    }
-    let domedir = homedir.unwrap().join(".wasmdome/");
-    let _ = std::fs::create_dir_all(&domedir)?;
-    if let CredentialsResponse::Valid {
-        user_jwt,
-        user_secret,
-    } = creds
-    {
-        let fcontents = format!(
-            r#"
------BEGIN NATS USER JWT-----
-{}
-------END NATS USER JWT------
-
-************************* IMPORTANT *************************
-NKEY Seed printed below can be used to sign and prove identity.
-NKEYs are sensitive and should be treated as secrets.
-
------BEGIN USER NKEY SEED-----
-{}
-------END USER NKEY SEED------
-
-*************************************************************
-"#,
-            user_jwt, user_secret
+fn run_match(
+    nc: nats::Connection,
+    max_turns: u32,
+    board_height: u32,
+    board_width: u32,
+) -> Result<(), Box<dyn Error>> {
+    let match_id = Uuid::new_v4().to_string();
+    let sub = nc.subscribe("wasmdome.public.arena.events")?;
+
+    let req = nc.request_timeout(
+        "wasmdome.internal.arena.control",
+        &serde_json::to_vec(&QueryMechs)?,
+        std::time::Duration::from_millis(1500),
+    );
+
+    if req.is_err() {
+        println!(
+            "No response from engine-provider, please ensure you have an engine-provider running."
+        );
+        return Ok(());
+    };
+
+    let res: MechQueryResponse = serde_json::from_slice(&req.unwrap().data)?;
+
+    if res.mechs.len() < 1 {
+        println!(
+            "No mechs were found in the lattice. Ensure you have scheduled at least one mech."
         );
-        let domefile = domedir.join("arena.creds");
-        std::fs::write(domefile, fcontents.as_bytes())?;
-        println!("New arena credentials written to ~/.wasmdome/arena.creds");
+        return Ok(());
+    };
+
+    let cm = StartMatch(CreateMatch {
+        match_id: match_id.clone(),
+        actors: res.mechs.iter().map(|m| m.id.clone()).collect(),
+        board_height,
+        board_width,
+        max_turns,
+        aps_per_turn: domain::state::APS_PER_TURN,
+    });
+
+    nc.request_timeout(
+        "wasmdome.internal.arena.control",
+        &serde_json::to_vec(&cm)?,
+        std::time::Duration::from_millis(500),
+    )?;
+
+    let (s, r) = unbounded();
+
+    sub.with_handler(move |msg| {
+        let msg = serde_json::from_slice(&msg.data).unwrap();
+        match msg {
+            Some(ArenaEvent::MatchCompleted {
+                match_id, cause, ..
+            }) => {
+                s.send(format!(
+                    "Match \"{}\" completed.\nCause: {:?}",
+                    match_id, cause
+                ))
+                .unwrap();
+            }
+            _ => (),
+        };
         Ok(())
-    } else {
-        Err("Did not obtain valid arena credentials".into())
-    }
+    });
+
+    println!("{}", r.recv().unwrap());
```

Comment:
Can we do this where you're using a `select!` to receive either the value on the channel or a timeout period? This will prevent the CLI from hanging indefinitely if the match completed event never arrives and we can display a message indicating that's what happened.

Reply:
I added a 5s timeout and an error message, if either of these values are inappropriate let me know what you think

Reply:
5s might be a bit long but it's fine, as long as there's some cap on it. I'd totally forgotten about recv_timeout. Nice!